File: apps/worker_whatsmeow/internal/app/schedule_status_identity.go
package app

import (
	"crypto/sha256"
	"encoding/hex"
	"strings"
)

const scheduleStatusEventIDVersion = "v1"

func ensureScheduleStatusEventID(update *ScheduleStatusUpdate) string {
	if update == nil {
		return ""
	}
	if existing := strings.TrimSpace(update.EventID); existing != "" {
		update.EventID = existing
		return existing
	}
	canonicalParts := []string{
		scheduleStatusEventIDVersion,
		strings.TrimSpace(update.ScheduleID),
		strings.TrimSpace(update.ContactID),
		strings.TrimSpace(update.MessageID),
		strings.TrimSpace(update.Status),
	}
	canonical := strings.Join(canonicalParts, "\x00")
	digest := sha256.Sum256([]byte(canonical))
	update.EventID = "schedule_status_" + scheduleStatusEventIDVersion + "_" + hex.EncodeToString(digest[:])
	return update.EventID
}

File: packages/common/functions/scheduleStatusIdentity.ts
import { createHash } from 'node:crypto';
import { IScheduleStatusUpdate } from '../interfaces/IScheduleStatusUpdate';

const SCHEDULE_STATUS_EVENT_ID_VERSION = 'v1';

function normalize(value: string): string {
  return value.trim();
}

export function buildScheduleStatusKafkaKey(
  data: Pick<IScheduleStatusUpdate, 'schedule_id' | 'contact_id' | 'message_id'>
): string {
  return [data.schedule_id, data.contact_id, data.message_id]
    .map(normalize)
    .join(':');
}

export function buildScheduleStatusEventId(
  data: Pick<
    IScheduleStatusUpdate,
    'schedule_id' | 'contact_id' | 'message_id' | 'status'
  >
): string {
  const canonical = [
    SCHEDULE_STATUS_EVENT_ID_VERSION,
    normalize(data.schedule_id),
    normalize(data.contact_id),
    normalize(data.message_id),
    data.status,
  ].join('\0');

  return `schedule_status_${SCHEDULE_STATUS_EVENT_ID_VERSION}_${createHash(
    'sha256'
  )
    .update(canonical)
    .digest('hex')}`;
}

export function ensureScheduleStatusEventId(
  data: IScheduleStatusUpdate
): string {
  const existing = data.event_id?.trim();
  if (existing) {
    data.event_id = existing;
    return existing;
  }

  const generated = buildScheduleStatusEventId(data);
  data.event_id = generated;
  return generated;
}
